Parser4 merge

fmouronh opened the pull request:
Classes ConfigParser, ServerConfig and ServerLocation (+ derived) almost fully implemented
Missing features:
* handling edge cases in getter functions (eg, index out of bounds)
* derived class from ServerLocation to handle CGI location blocks
* configurations for file uploads
* validation of loaded configurations
* syntax check for config file (this will be done in a separate static class)

Because the current implementation only does a partial check for the validity of the configs provided, please make sure that there are no errors in the config file provided when running webserv; otherwise, unspecified behaviour may occur

fmouronh committed:
file contents are now being loaded into vector

fmouronh committed:
implementing method to override default values for root and index directives

fmouronh committed:
implemented override for default root and index directives; bug fixes

fmouronh committed:
cleaning up for merge

fmouronh committed:
Merge branch 'main' into Parser4Merge

fmouronh committed:
trying to fix circular include hell with LOCATION_BLOCK_TYPE typedef enum

fmouronh committed:
preparing for next merge with main; fixed compilation errors

fmouronh committed:
segfaults fixed

fmouronh committed:
cleaning up

fmouronh committed:
preparing for next merge

fmouronh committed:
cleaning up

MrFacundo commented:
- Program not running because of out of bounds issue with the Utils::sTrim(). This fixed the issue for me

`	std::string sTrim(std::string const& str) {
		if (str.empty()) return str;

		std::string::size_type start = 0;
		std::string::size_type end = str.size();

		while (start < end && std::isspace(str[start])) {
			start++;
		}

		while (end > start && std::isspace(str[end - 1])) {
			end--;
		}

		return str.substr(start, end - start);
	}
`


- There seems to be an issue in ConfigParser::parseRedirections

where the redirection strings are not trimmed of whitespaces:

"originalpath.html" " newpath.html", which causes issues on the redirection implementation. It would be nice to solve the issue at the parsing phase.

fmouronh commented:
Added the conditional to Utils::sTrim to check if the string is empty.

I noticed the bug in ConfigParser::parseRedirections while working on the SyntaxChecker class and fixed it right away. All fixes will be present in the next PR.

You can ignore this one, as the next PR will be done within 24hours
